go/vt/vtgate/endtoend/vstream_test.go

```diff
@@ -232,6 +233,129 @@ func TestVStreamCopyBasic(t *testing.T) {
+
+	_, err := conn.ExecuteFetch("insert into t1(id1,id2) values(1,1), (2,2), (3,3), (4,4)", 1, false)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	reader, err := gconn.VStream(ctx, topodatapb.TabletType_PRIMARY, vgtid, filter, flags)
+	if err != nil {
+		t.Fatal(err)
+	}
+	require.NotNil(t, reader)
+	expectedRowCopyEvents := 4 // id1 and id2 IN(1,2,3,4)
+	expectedCatchupEvents := 0
+	expectedEvents := []string{
+		`type:ROW row_event:{table_name:"ks.t1" row_changes:{after:{lengths:1 lengths:1 values:"11"}} keyspace:"ks" shard:"-80"} keyspace:"ks" shard:"-80"`,
+		`type:ROW row_event:{table_name:"ks.t1" row_changes:{after:{lengths:1 lengths:1 values:"22"}} keyspace:"ks" shard:"-80"} keyspace:"ks" shard:"-80"`,
+		`type:ROW row_event:{table_name:"ks.t1" row_changes:{after:{lengths:1 lengths:1 values:"33"}} keyspace:"ks" shard:"-80"} keyspace:"ks" shard:"-80"`,
+		`type:ROW row_event:{table_name:"ks.t1" row_changes:{after:{lengths:1 lengths:1 values:"44"}} keyspace:"ks" shard:"80-"} keyspace:"ks" shard:"80-"`,
+	}
+	confirmVStreamEvents(t, cancel, reader, expectedEvents, expectedRowCopyEvents, expectedCatchupEvents)
+
+	// close the connection and the vstream with it
+	gconn.Close()
+
+	// Now we want to pick up where we left off, so we get the current position and set
+	// the lastPK to what it was, id1=4, so we should only copy rows for id1 IN(5,6,7,8,9)
+	mpos, err = mconn.PrimaryPosition()
+	if err != nil {
+		t.Fatal(err)
+	}
+	lastPK := sqltypes.Result{
+		Fields: []*query.Field{{Name: "id1", Type: query.Type_INT64}},
+		Rows:   [][]sqltypes.Value{{sqltypes.NewInt64(4)}},
+	}
+	tableLastPK := []*binlogdatapb.TableLastPK{{
+		TableName: "t1",
+		Lastpk:    sqltypes.ResultToProto3(&lastPK),
+	}}
+	shardGtids = []*binlogdatapb.ShardGtid{}
+	shardGtids = append(shardGtids, &binlogdatapb.ShardGtid{
+		Keyspace: "ks",
+		Shard:    "-80",
+		Gtid:     fmt.Sprintf("%s/%s", mpos.GTIDSet.Flavor(), mpos),
+		TablePKs: tableLastPK,
+	})
+	shardGtids = append(shardGtids, &binlogdatapb.ShardGtid{
+		Keyspace: "ks",
+		Shard:    "80-",
+		Gtid:     fmt.Sprintf("%s/%s", mpos.GTIDSet.Flavor(), mpos),
+		TablePKs: tableLastPK,
+	})
+	vgtid.ShardGtids = shardGtids
+
+	// Open a new connection (as we closed the original)
+	gconn, err = vtgateconn.Dial(ctx, grpcAddress)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer gconn.Close()
+
+	reader, err = gconn.VStream(ctx, topodatapb.TabletType_PRIMARY, vgtid, filter, flags)
+	if err != nil {
+		t.Fatal(err)
+	}
+	require.NotNil(t, reader)
+
+	_, err = conn.ExecuteFetch("insert into t1(id1,id2) values (5,5), (6,6), (7,7), (8,8), (9,9)", 1, false)
+	if err != nil {
+		t.Fatal(err)
+	}
+	_, err = conn.ExecuteFetch("update t1 set id2 = 119 where id1 = 2", 1, false)
+	if err != nil {
+		t.Fatal(err)
+	}
+	_, err = conn.ExecuteFetch("update t1 set id2 = 219 where id1 = 7", 1, false)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	expectedRowCopyEvents = 5 // id1 and id2 IN(5,6,7,8,9)
+	expectedCatchupEvents = 2
```

Comment:
Actually, the two update events will be caught in the replication phase, not the initial copy/catchup phase.
The expected catchup is [skipped](https://github.com/vitessio/vitess/blob/567fe98ce6e54db2324733e58ce5fc0bfc18fba3/go/vt/vttablet/tabletserver/vstreamer/copy.go#L51) because of missing `uvs.pos`.

The test is passed because the updates happened after the vstream started the copy.

Reply:
I want to pass the test changed like the below.
It means that the next connection can fetch all the changes in between.

```diff
diff --git a/go/vt/vtgate/endtoend/vstream_test.go b/go/vt/vtgate/endtoend/vstream_test.go
index a523d5e1eb..f7ce55c8b0 100644
--- a/go/vt/vtgate/endtoend/vstream_test.go
+++ b/go/vt/vtgate/endtoend/vstream_test.go
@@ -323,12 +323,6 @@ func TestVStreamCopyResume(t *testing.T) {
        }
        defer gconn.Close()

-       reader, err = gconn.VStream(ctx, topodatapb.TabletType_PRIMARY, vgtid, filter, flags)
-       if err != nil {
-               t.Fatal(err)
-       }
-       require.NotNil(t, reader)
-
        _, err = conn.ExecuteFetch("insert into t1(id1,id2) values (5,5), (6,6), (7,7), (8,8), (9,9)", 1, false)
        if err != nil {
                t.Fatal(err)
@@ -342,6 +336,12 @@ func TestVStreamCopyResume(t *testing.T) {
                t.Fatal(err)
        }

+       reader, err = gconn.VStream(ctx, topodatapb.TabletType_PRIMARY, vgtid, filter, flags)
+       if err != nil {
+               t.Fatal(err)
+       }
+       require.NotNil(t, reader)
+
        expectedRowCopyEvents = 5 // id1 and id2 IN(5,6,7,8,9)
        expectedCatchupEvents = 2
        expectedEvents = []string{
```

Reply:
From this point of view, I thought [the previous conditional](https://github.com/vitessio/vitess/pull/11103/commits/9971669e28b63e6c8fecdea106dd07c935cf1844#diff-32dbe2effb6b8b7e1151736ee190783bb236feb90ad7c04fe8ec6eec3ebea30fL360) was necessary.